harmoni-web-api/src/main/java/org/harmoniapp/harmoniwebapi/services/AiScheduleService.java

```diff
@@ -0,0 +1,310 @@
+                .flatMap(reqShiftDto -> reqShiftDto.roles().stream())
+                .map(ReqRoleDto::roleId)
+                .collect(Collectors.toSet());
+
+        List<Employee> employees = new ArrayList<>();
+        for (User user : users) {
+            Set<Role> userRoles = user.getRoles();
+            for (Role role : userRoles) {
+                if (validRoles.contains(role.getId())) {
+                    Employee employee = new Employee(user.getEmployeeId(), role.getName());
+                    employees.add(employee);
+                    break;
+                }
+            }
+        }
+        return employees.stream().collect(Collectors.groupingBy(Employee::role));
+    }
+
+    /**
+     * Prepares shifts for the schedule generation.
+     *
+     * @param scheduleRequirements the list of schedule requirements
+     * @return a list of shifts
+     */
+    private List<Shift> prepareShifts(List<ScheduleRequirement> scheduleRequirements, List<PredefineShift> predefineShifts,
+                                      List<Role> roles) {
+        List<Shift> shifts = new ArrayList<>();
+
+        for (ScheduleRequirement scheduleRequirement : scheduleRequirements) {
+
+            scheduleRequirement.shifts()
+                    .sort(Comparator.comparing(rs -> predefineShifts.stream()
+                            .filter(ps -> ps.getId().equals(rs.shiftId()))
+                            .findFirst()
+                            .orElseThrow()
+                            .getStart())
+                    );
+            for (ReqShiftDto reqShiftDto : scheduleRequirement.shifts()) {
+                List<Requirements> requirements = prepareRequirements(reqShiftDto.roles(), roles);
+                shifts.add(new Shift(reqShiftDto.shiftId().intValue(),
+                        scheduleRequirement.date().getDayOfYear(),
+                        requirements));
+            }
+        }
+        return shifts;
+    }
+
+    /**
+     * Prepares requirements for the schedule generation.
+     *
+     * @param requirements the list of requirements
+     * @return a list of requirements
+     */
+    private List<Requirements> prepareRequirements(List<ReqRoleDto> requirements, List<Role> roles) {
+        List<Requirements> req = new ArrayList<>(requirements.size());
+        for (ReqRoleDto reqRoleDto : requirements) {
+            Role role = roles.stream().filter(r -> Objects.equals(r.getId(), reqRoleDto.roleId())).findFirst().orElseThrow();
+            req.add(new Requirements(role.getName(), reqRoleDto.quantity()));
+        }
+        return req;
+    }
+
+    /**
+     * Decodes shifts from shift representation in algorithm to entity.
+     *
+     * @param shifts the list of shifts
+     * @return a list of decoded shifts
+     */
+    private List<org.harmoniapp.harmonidata.entities.Shift> decodeShifts(List<Shift> shifts, List<User> users,
```

Comment:
List<org.harmoniapp.harmonidata.entities.Shift> I think we can replace it with more simple import